src/runtime/agas/server/symbol_namespace_server.cpp

```diff
@@ -366,31 +370,57 @@ naming::gid_type symbol_namespace::unbind(std::string const& key)
 } // }}}
 
 // TODO: catch exceptions
-void symbol_namespace::iterate(
-    symbol_namespace::iterate_names_function_type const& f
-    )
+symbol_namespace::iterate_names_return_type symbol_namespace::iterate(
+    std::string const& pattern)
 { // {{{ iterate implementation
     util::scoped_timer<std::atomic<std::int64_t> > update(
         counter_data_.iterate_names_.time_
     );
     counter_data_.increment_iterate_names_count();
-    std::unique_lock<mutex_type> l(mutex_);
 
-    for (gid_table_type::iterator it = gids_.begin(); it != gids_.end(); ++it)
+    std::map<std::string, naming::gid_type> found;
+
+    if (pattern.find_first_of("*?[]") != std::string::npos)
     {
-        std::string key(it->first);
-        naming::gid_type gid = *(it->second);
+        std::string str_rx(util::regex_from_pattern(pattern, throws));
+        boost::regex rx(str_rx, boost::regex::perl);
 
+        std::unique_lock<mutex_type> l(mutex_);
+        for (gid_table_type::iterator it = gids_.begin(); it != gids_.end();
+             ++it)
         {
+            if (!boost::regex_match(it->first, rx))
+                continue;
+
+            // hold on to entry while map is unlocked
+            std::shared_ptr<naming::gid_type> current_gid(it->second);
             util::unlock_guard<std::unique_lock<mutex_type> > ul(l);
-            f(key, gid);
+
+            found[it->first] =
+                naming::detail::split_gid_if_needed(*current_gid).get();
```

Comment:
Yeah, this will work. While going through the PR I was really wondering why we don't reuse `hpx::naming::id_type` here instead of `std::shared_ptr<gid_type>`. But that's unrelated to that PR.

Reply:
AGAS stores `gid_type` instances only by design. For all intents and purposes this is sufficient and avoids problems with reference counting etc. The symbol namespace is special in that regard that it actually has to maintain reference counts (for some of its functionality). I'm impartial to changing this - it wouldn't make the code so much more readable, however.